src/dlr_tensorflow/dlr_tensorflow.cc

```diff
@@ -0,0 +1,290 @@
+  ListDir(dirname, paths_vec);
+  for (auto filename : paths_vec) {
+    std::string basename = GetBasename(filename);
+    if (EndsWith(filename, ".pb")) {
+      if (pb_file.empty()) {
+        pb_file = filename;
+      } else {
+        LOG(FATAL) << "Multiple .pb files under the folder: " << dirname;
+      }
+    }
+  }
+  if (pb_file.empty()) {
+    LOG(FATAL) << "No Tensorflow frozen model file found under folder: " << dirname;
+  }
+  return pb_file;
+}
+
+void dlr::FreeBuffer(void* data, size_t length) {
+  free(data);
+}
+
+TF_Buffer* dlr::ReadTFFile(const char* file) {
+  FILE *f = fopen(file, "rb");
+  fseek(f, 0L, SEEK_END);
+  long fsize = ftell(f);
+  fseek(f, 0L, SEEK_SET);  //same as rewind(f);
+
+  void* data = malloc(fsize);
+  if (fread(data, fsize, 1, f) != 1) {
+    printf("File read error....\n");
+  }
+  fclose(f);
+
+  TF_Buffer* buf = TF_NewBuffer();
+  buf->data = data;
+  buf->length = fsize;
+  buf->data_deallocator = FreeBuffer;
+  return buf;
+}
+
+void TensorflowModel::LoadFrozenModel(const char* pb_file) {
+  const char* op_prefix = "";
+  TF_ImportGraphDefOptions* opts = TF_NewImportGraphDefOptions();
+  TF_ImportGraphDefOptionsSetPrefix(opts, op_prefix);
+
+  TF_Buffer* graph_def = ReadTFFile(pb_file);
+  TF_GraphImportGraphDef(graph_, graph_def, opts, status_);
+  if (TF_GetCode(status_) != TF_OK) {
+    LOG(FATAL) << "ERROR: Unable to import graph " << TF_Message(status_);
+    return; // unreachable
+  }
+  LOG(INFO) << "Successfully imported graph";
+  TF_DeleteImportGraphDefOptions(opts);
+  TF_DeleteBuffer(graph_def);
+}
+
+void TensorflowModel::GenTensorSpec(bool isInput) {
+  std::vector<std::string> tensor_names = isInput ? input_names_ : output_names_;
+  std::regex r("^(.+):(\\d+)$");
+  for (std::string t_name : tensor_names) {
+    std::smatch match;
+    std::string op_name;
+    int op_out_id;
+    if (std::regex_search(t_name, match, r)) {
+      op_name = match.str(1);
+      op_out_id = std::stoi(match.str(2));
+    } else {
+      LOG(FATAL) << "ERROR: failed to parse tensor name " << t_name;
+      return; // unreachable
+    }
+
+    TF_Operation* op = TF_GraphOperationByName(graph_, op_name.c_str());
+    if (op == NULL) {
+      LOG(FATAL) << "ERROR: inputOp TF_GraphOperationByName failed for operation " << op_name;
+      return; // unreachable
+    }
+    TF_Output oper_out = {op, op_out_id};
+    int n_dim = TF_GraphGetTensorNumDims(graph_, oper_out, status_);
+    if (TF_GetCode(status_) != TF_OK) {
+      LOG(FATAL) << "ERROR: TF_GraphGetTensorNumDims failed " << TF_Message(status_);
+      return; // unreachable
+    }
+    int64_t dims[n_dim];
+    TF_GraphGetTensorShape(graph_, oper_out, dims, n_dim, status_);
+    if (TF_GetCode(status_) != TF_OK) {
+      LOG(FATAL) << "ERROR: TF_GraphGetTensorShape failed " << TF_Message(status_);
+      return; // unreachable
+    }
+    if (dims[0] == -1) {
+      dims[0] = 1;
+    }
+    size_t num_pixels = dims[0];
```

Comment:
Should we use `int64_t` instead of `size_t` since `size_t` could be 32 bits?

Reply:
Lets use `num_elements` instead of `num_pixels` since we aren't always using images.

Reply:
n_dim can be 0 when the tensor is a scalar. So `num_elements` should be initialized to 1 and the following loop should start at dim 0.

Scalars:
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/c_api_test.cc#L440-L447

Reply:
Since TF_AllocateTensor needs `size_t` for 'len` we decided to use `size_t` for `num_elements`
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/tf_tensor.h#L91